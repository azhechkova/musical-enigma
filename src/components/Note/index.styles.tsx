import { HTMLAttributes } from 'react';
import styled from 'styled-components';

interface BasicNoteProps extends HTMLAttributes<HTMLButtonElement> {
  isPressed?: boolean;
}

export const BasicNote = styled.button<BasicNoteProps>(({ isPressed }) => [
  `
  display: block;
  padding: 0;
  border-color: var(--white-hover);
  cursor: pointer;
  height: 200px;
  width: 50px;
  box-shadow: -1px 0 0 var(--white-hover) inset, 0 0 5px hsl(0, 0%, 80%) inset,
    0 0 3px var(--black);
  background: linear-gradient(to bottom, hsl(0, 0%, 93%) 0%, var(--white) 100%);

  &:first-child {
    border-radius: 4px 0 4px 4px;
  }
  &:last-child {
    border-radius: 0 4px 4px 4px;
  }
  &:active {
    border-top: 1px solid #777;
    border-left: 1px solid #999;
    border-bottom: 1px solid #999;
    box-shadow: 2px 0 3px rgba(0, 0, 0, 0.1) inset,
      -5px 5px 20px rgba(0, 0, 0, 0.2) inset, 0 0 3px rgba(0, 0, 0, 0.2);
    background: linear-gradient(to bottom, var(--white) 0%, #e9e9e9 100%);
  }
`,
  isPressed &&
    `border-top: 1px solid #777;
    border-left: 1px solid #999;
    border-bottom: 1px solid #999;
    box-shadow: 2px 0 3px rgba(0, 0, 0, 0.1) inset,
      -5px 5px 20px rgba(0, 0, 0, 0.2) inset, 0 0 3px rgba(0, 0, 0, 0.2);
    background: linear-gradient(to bottom, var(--white) 0%, #e9e9e9 100%);
`,
]);

export const BasicSemiNote = styled(BasicNote)(({ isPressed }) => [
  `
  width: 2em;
  margin: 0 -1em;
  z-index: 10;
  position: relative;
  height: 100px;
  box-shadow: -1px -1px 2px var(--black-hover) inset,
    0 -5px 2px 3px var(--black) inset, 0 2px 4px var(--black-hover);
  background: linear-gradient(45deg, hsl(0, 0%, 13%) 0%, hsl(0, 0%, 33%) 100%);

  &:active {
    box-shadow: -1px -1px 2px rgba(232, 237, 223, 0.2) inset,
      0 -2px 2px 3px rgba(36, 36, 35, 0.6) inset,
      0 1px 2px rgba(36, 36, 35, 0.5);
    background: linear-gradient(to right, #444 0%, #222 100%);
  }
`,
  isPressed &&
    `box-shadow: -1px -1px 2px rgba(232, 237, 223, 0.2) inset,
      0 -2px 2px 3px rgba(36, 36, 35, 0.6) inset,
      0 1px 2px rgba(36, 36, 35, 0.5);
    background: linear-gradient(to right, #444 0%, #222 100%);
`,
]);
