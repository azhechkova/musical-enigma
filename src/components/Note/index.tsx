import React from 'react';
import { BasicNote, BasicSemiNote } from './index.styles';

interface NotePress extends NoteType {
  isPressed?: boolean;
}

interface NoteProps {
  note: NotePress;
  playNote: (note: string) => void;
  stopPlay: () => void;
}

const Note = ({ note, playNote, stopPlay }: NoteProps): JSX.Element => {
  const onMouseDown = (): void => playNote(note.name);

  const onMouseUp = (e: React.MouseEvent<HTMLButtonElement>): void => {
    e.preventDefault();
    e.stopPropagation();
    stopPlay();
  };

  const NoteComponent = note.type === 'tone' ? BasicNote : BasicSemiNote;

  return (
    <NoteComponent
      onMouseLeave={onMouseUp}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      isPressed={note.isPressed}
      type="button"
      title={note.name}
    />
  );
};

export default Note;
