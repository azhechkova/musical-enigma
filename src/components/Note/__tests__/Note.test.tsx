import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import Note from '..';

const whiteNote = { key: 'KeyA', name: 'C4', type: 'tone' } as NoteType;
const blackNote = { key: 'KeyA', name: 'C4', type: 'tone' } as NoteType;
const noteProps = {
  playNote: jest.fn(),
  stopPlay: jest.fn(),
};

describe('Note', () => {
  it('render white button properly', () => {
    const { container: note } = render(
      <Note note={whiteNote} {...noteProps} />,
    );

    expect(note).toMatchSnapshot();
  });
  it('renders black button properly', () => {
    const { container: note } = render(
      <Note note={blackNote} {...noteProps} />,
    );

    expect(note).toMatchSnapshot();
  });
  it('has button with title', async () => {
    const container = render(<Note note={whiteNote} {...noteProps} />);
    const button = await container.findByTitle(whiteNote.name);

    expect(button).toBeInTheDocument();
  });
  it('calls playNote function', async () => {
    const container = render(<Note note={whiteNote} {...noteProps} />);
    const button = await container.findByTitle(whiteNote.name);

    expect(button).toBeInTheDocument();
    fireEvent.mouseDown(button);
    expect(noteProps.playNote).toHaveBeenCalledTimes(1);
  });
  it('calls stopPlay function', async () => {
    const container = render(<Note note={whiteNote} {...noteProps} />);
    const button = await container.findByTitle(whiteNote.name);

    expect(button).toBeInTheDocument();
    fireEvent.mouseDown(button);
    fireEvent.mouseUp(button);

    expect(noteProps.stopPlay).toHaveBeenCalledTimes(1);
  });
});
