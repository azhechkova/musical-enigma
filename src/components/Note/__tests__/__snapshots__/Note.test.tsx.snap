// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note render white button properly 1`] = `
.c0 {
  display: block;
  padding: 0;
  border-color: var(--white-hover);
  cursor: pointer;
  height: 200px;
  width: 50px;
  box-shadow: -1px 0 0 var(--white-hover) inset,0 0 5px hsl(0,0%,80%) inset, 0 0 3px var(--black);
  background: linear-gradient(to bottom,hsl(0,0%,93%) 0%,var(--white) 100%);
}

.c0:first-child {
  border-radius: 4px 0 4px 4px;
}

.c0:last-child {
  border-radius: 0 4px 4px 4px;
}

.c0:active {
  border-top: 1px solid #777;
  border-left: 1px solid #999;
  border-bottom: 1px solid #999;
  box-shadow: 2px 0 3px rgba(0,0,0,0.1) inset, -5px 5px 20px rgba(0,0,0,0.2) inset,0 0 3px rgba(0,0,0,0.2);
  background: linear-gradient(to bottom,var(--white) 0%,#e9e9e9 100%);
}

<div>
  <button
    class="c0"
    title="C4"
    type="button"
  />
</div>
`;

exports[`Note renders black button properly 1`] = `
.c0 {
  display: block;
  padding: 0;
  border-color: var(--white-hover);
  cursor: pointer;
  height: 200px;
  width: 50px;
  box-shadow: -1px 0 0 var(--white-hover) inset,0 0 5px hsl(0,0%,80%) inset, 0 0 3px var(--black);
  background: linear-gradient(to bottom,hsl(0,0%,93%) 0%,var(--white) 100%);
}

.c0:first-child {
  border-radius: 4px 0 4px 4px;
}

.c0:last-child {
  border-radius: 0 4px 4px 4px;
}

.c0:active {
  border-top: 1px solid #777;
  border-left: 1px solid #999;
  border-bottom: 1px solid #999;
  box-shadow: 2px 0 3px rgba(0,0,0,0.1) inset, -5px 5px 20px rgba(0,0,0,0.2) inset,0 0 3px rgba(0,0,0,0.2);
  background: linear-gradient(to bottom,var(--white) 0%,#e9e9e9 100%);
}

<div>
  <button
    class="c0"
    title="C4"
    type="button"
  />
</div>
`;
